pkgs/development/compilers/gcc/common/configure-flags.nix

```diff
@@ -244,6 +244,9 @@ let
     ++ lib.optionals (langD) [
       "--with-target-system-zlib=yes"
     ]
+    ++ lib.optionals (targetPlatform.system == "x86_64-darwin") [
```

Comment:
That condition should probably be phrased as `targetPlatform.isDarwin && targetPlatform.isx86_64`. Why is it `x86_64`-specific? Shouldn't aarch64 have the same problem?

Reply:
It’s an Intel-specific issue. The Darwin stdenv rework switched the assembler on x86_64-darwin to the clang integrated assembler. The intent was to match what the platform tools does (Apple has been using the clang assembler since Xcode 12, which uses LLVM 10). It appears there is an assembly optimization that is sensitive to whether there is debug info in the binary. In this case, GCC builds stage 2 with debug info and stage 3 without.

```diff
-stage2-libiberty/sort.o:
+stage3-libiberty/sort.o:
 (__TEXT,__text) section
 _sort_pointers:
 0000000000000000	55              	pushq	%rbp                            ## Latency: 5
@@ -44,8 +44,8 @@
 000000000000009b	4a8d3421        	leaq	_sort_pointers(%rcx,%r12), %rsi
 000000000000009f	4c01fe          	addq	%r15, %rsi
 00000000000000a2	4839f0          	cmpq	%rsi, %rax
-00000000000000a5	731d            	jae	0xc4
-00000000000000a7	660f1f840000000000	nopw	_sort_pointers(%rax,%rax)
+00000000000000a5	0f8319000000    	jae	0xc4
+00000000000000ab	0f1f440000      	nopl	(%rax,%rax)
 00000000000000b0	0fb610          	movzbl	_sort_pointers(%rax), %edx      ## Latency: 5
 00000000000000b3	4883c008        	addq	$0x8, %rax
 00000000000000b7	838495d0fbffff01	addl	$0x1, -0x430(%rbp,%rdx,4)       ## Latency: 7
@@ -59,11 +59,11 @@
 00000000000000de	4883c004        	addq	$0x4, %rax
 00000000000000e2	8950fc          	movl	%edx, -0x4(%rax)
 00000000000000e5	4839f0          	cmpq	%rsi, %rax
-00000000000000e8	75ee            	jne	0xd8
-00000000000000ea	488b95c0fbffff  	movq	-0x440(%rbp), %rdx              ## Latency: 5
-00000000000000f1	4c01fa          	addq	%r15, %rdx
-00000000000000f4	7231            	jb	0x127
-00000000000000f6	662e0f1f840000000000	nopw	%cs:_sort_pointers(%rax,%rax)
+00000000000000e8	0f85eaffffff    	jne	0xd8
+00000000000000ee	488b95c0fbffff  	movq	-0x440(%rbp), %rdx              ## Latency: 5
+00000000000000f5	4c01fa          	addq	%r15, %rdx
+00000000000000f8	0f8229000000    	jb	0x127
+00000000000000fe	6690            	nop
 0000000000000100	420fb60c22      	movzbl	_sort_pointers(%rdx,%r12), %ecx ## Latency: 5
 0000000000000105	4883ea08        	subq	$0x8, %rdx
 0000000000000109	8b848dd0fbffff  	movl	-0x430(%rbp,%rcx,4), %eax       ## Latency: 5
```

When I use a clang 16 stdenv, its stage 3 disassembly matches the stage 2 above. I’m running builds with clang 12, 13, 14, and 15 to confirm when the issue is fixed; but I’m pretty sure it will be clang 12. https://github.com/llvm/llvm-project/commit/a048ce13e32daa255d26533c00da8abd0b67e819 disabled the optimization (even applied to assembly files) by default, and it is tagged for LLVM 12.

If that’s the case, I would rather push a fix to cctools-llvm them have GCC explicitly use cctools-port. If this issue is affecting GCC, it could also affect other applications or user code with debug information. It doesn’t seem to be a correctness issue, but it affects code generation in an apparently unwanted way.

Reply:
Ideally `gcc` should use the same assembler in `stage2` and `stage3`. Do I understand correctly it uses different ones? That would be very scary.

Reply:
Oh, I did not read your comment carefully. Looks like it's just an assembler bug then. Worth adding an explicit comment into the source code that it's a workaround for a specific bug?

Reply:
I’ve opened #242202 with some cctools-llvm fixes including one for this issue. I targeted staging-next due to the GCC build failure, but I can retarget staging instead if it would be preferred to use this in the interim.

The fix in #242202 is intended to be more general since the problem only affects LLVM 11. Once #242094 is merged, Darwin will use LLVM 16 and not have this issue with the integrated assembler.